### PR TITLE
Bug 1878591:  Added tests for lso ocs disk inventory

### PR DESCRIPTION
**Pull request by afreen23**
Screenshot from 2020-09-13 17-35-17
Signed-off-by: Afreen Rahman <afrahman@redhat.com>

### CODE REVIEW

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        `kubectl get localvolumediscoveryresults -n ${LOCAL_STORAGE_NAMESPACE} -o json`,
+      ).toString(),
+    );
+    const lvdr: LocalVolumeDiscoveryResultKind = lvdrJson.items[0];
+    const { nodeName } = lvdr.spec;
+    discoveredDevices = lvdr.status.discoveredDevices;
+    await browser.get(`${appHost}/`);
+    await clickNavLink(['Compute', 'Nodes']);
+    await isLoaded();
+    $(`a[data-test-id="${nodeName}"]`).click();
```
Comment (gnehapk) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:35]: All the selectors should be moved to a corresponding view.ts file, similar to other tests.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  clearDiskFilter,
+} from '../views/disks-list-page.views';
+import { LOCAL_STORAGE_NAMESPACE } from '../../src/constants/index';
+import {
+  LocalVolumeDiscoveryResultKind,
+  DiskMetadata,
+  DiskStates,
+} from '../../src/components/disks-list/types';
+
+// Pre-requirement: Local volume discovery to be present
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:20]:
```suggestion
// Prerequisite: Local volume discovery to be present
```

File: frontend/packages/local-storage-operator-plugin/src/components/disks-list/disks-list-page.tsx [hunk 3]
```diff
@@ -97,9 +97,7 @@ const diskRow: RowFunction<DiskMetadata> = ({ obj, index, key, style }) => (
     <TableData className={tableColumnClasses[1]}>{obj.status.state}</TableData>
     <TableData className={tableColumnClasses[2]}>{obj.type || '-'}</TableData>
     <TableData className={cx(tableColumnClasses[3], 'co-break-word')}>{obj.model || '-'}</TableData>
-    <TableData className={tableColumnClasses[4]}>
-      {humanizeBinaryBytes(obj.size).string || '-'}
-    </TableData>
+    <TableData className={tableColumnClasses[4]}>{humanizeBinaryBytes(obj.size).string}</TableData>
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/src/components/disks-list/disks-list-page.tsx:100]: Why this change?

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    const lvdr: LocalVolumeDiscoveryResultKind = lvdrJson.items[0];
+    const { nodeName } = lvdr.spec;
+    discoveredDevices = lvdr.status.discoveredDevices;
+    await browser.get(`${appHost}/`);
+    await clickNavLink(['Compute', 'Nodes']);
+    await isLoaded();
+    $(`a[data-test-id="${nodeName}"]`).click();
+  });
+  it('`Disks` tab is enabled and accessible', async () => {
+    await browser.wait(until.presenceOf(page.diskTab), 5000);
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:38]:
```suggestion
    await browser.wait(until.presenceOf(page.diskTab), 5 * SECOND);
```

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    page.diskTab.click();
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:39]:
```suggestion
    click(page.diskTab);
```
Rather than clicking a better test would be to expect an `elementToBeClickable`.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+} from '../views/disks-list-page.views';
+import { LOCAL_STORAGE_NAMESPACE } from '../../src/constants/index';
+import {
+  LocalVolumeDiscoveryResultKind,
+  DiskMetadata,
+  DiskStates,
+} from '../../src/components/disks-list/types';
+
+// Pre-requirement: Local volume discovery to be present
+describe('Disk List is accessible from Nodes view', () => {
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:21]:
```suggestion
describe('Disk list is accessible from Nodes view', () => {
```

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 1 above)
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:35]:
```suggestion
    click($(`a[data-test-id="${nodeName}"]`));
```

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  });
+  it('Rows are displayed with correct data format', async () => {
+    const rows = resourceRows;
+    expect(rows.count()).toBe(discoveredDevices.length);
+    const firstRow = rows.first().$$('td');
+    const firstDisk = discoveredDevices[0];
+    expect(firstRow.get(0).getText()).toBe(firstDisk.path);
+    expect(firstRow.get(1).getText()).toBe(firstDisk.status.state);
+    expect(firstRow.get(2).getAttribute('innerText')).toBe(firstDisk.type || '-');
+    expect(firstRow.get(3).getAttribute('innerText')).toBe(firstDisk.model || '-');
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:59]: `getText` also returns the visible inner text of the element. I am not sure why we would want to do it this way.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+export const clearDiskFilter = async () => {
+  const uncheckFilter = await $('.pf-c-chip-group__close >  button');
+  await uncheckFilter.click();
+};
+export const checkDiskFilter = async (filter: string) => {
+  await clickFilterDropdown();
+  const countBadge = $(`a[data-test-row-filter="${filter}"] > div > .pf-c-badge`);
+  const count = await countBadge.getText();
+  await $(`#${filter}`).click(); // clicks the filter in dropdown
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:26]:
```suggestion
  await click($(`#${filter}`));
```

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  dropdownFilter: $('button[data-test-id="filter-dropdown-toggle"]'),
+  dropdownFilterHeader: $('.pf-c-dropdown__group-title'),
+  textFilter: $('input[data-test-id="item-filter"]'),
+};
+
+export const clickFilterDropdown = () => diskList.dropdownFilter.click();
+
+export const clearDiskFilter = async () => {
+  const uncheckFilter = await $('.pf-c-chip-group__close >  button');
+  await uncheckFilter.click();
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:20]:
```suggestion
  await click(uncheckFilter);
```

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 1 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:35]: This takes in a dynamic value therefore I avoided creating anon functions.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 7 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:59]: `getAttribute` returns text for hidden elements and on certain screens few of the columns will be hidden

Comment (bipuladh) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:59]: since it will be running headless if `getText` works on headless mode then let us keep it as `getText`

Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:60]: No, `getText` , does not works therefore we are using `getAttribute` here.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import { click } from '@console/shared/src/test-utils/utils';
+
+export const page = {
+  diskTab: $('a[data-test-id="horizontal-link-Disks"]'),
+  resourceTitles: $$('[data-test-id="resource-title"]'),
+};
+
+export const diskList = {
+  headers: ['Name', 'Disk State', 'Type', 'Model', 'Capacity', 'Filesystem'],
+  dropdownFilter: $('button[data-test-id="filter-dropdown-toggle"]'),
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:12]: This selector is already present in the common selectors. Please import that.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  dropdownFilterHeader: $('.pf-c-dropdown__group-title'),
+  textFilter: $('input[data-test-id="item-filter"]'),
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:14]: Please import from crud views.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        `kubectl get localvolumediscoveryresults -n ${LOCAL_STORAGE_NAMESPACE} -o json`,
+      ).toString(),
+    );
+    const lvdr: LocalVolumeDiscoveryResultKind = lvdrJson.items[0];
+    const { nodeName } = lvdr.spec;
+    discoveredDevices = lvdr.status.discoveredDevices;
+    await browser.get(`${appHost}/`);
+    await clickNavLink(['Compute', 'Nodes']);
+    await isLoaded();
+    click($(`a[data-test-id="${nodeName}"]`));
```
Comment (gnehapk) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:36]: You can create a function and pass this dynamic value in view.ts file. 
It will be useful, if the selectors get changed in future version.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    /* @TODO: (afreen): importing `units.js` is creating a cyclic dependency. Fix the imports and modify test condition */
+    expect(firstRow.get(4).getAttribute('innerText')).not.toBe('-');
+    expect(firstRow.get(5).getAttribute('innerText')).toBe(firstDisk.fstype || '-');
+  });
+  it('Text filter is accessible and working', async () => {
+    const textbox = await diskList.textFilter;
+    await textbox.sendKeys(discoveredDevices[0].path);
+    let updatedRows = resourceRows.count();
+    expect(updatedRows).toBe(1);
+    // info: clear() is not triggering event to clear the filter and update rows
```
Comment (gnehapk) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:70]: I don't see clear method being called here. So I guess this comment can be removed.

Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:70]: https://github.com/openshift/console/pull/6606#discussion_r487679934

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 12 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:36]: https://github.com/openshift/console/pull/6606#discussion_r487679934

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 13 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:70]: See the lines 71-72, its for that and its valid here.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export const clickFilterDropdown = () => diskList.dropdownFilter.click();
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:17]:
```suggestion
export const clickFilterDropdown = () => click(diskList.dropdownFilter);
```

Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:17]: This function can go inside `diskList` object as well.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  await click(uncheckFilter);
+};
+export const checkDiskFilter = async (filter: string) => {
+  await clickFilterDropdown();
+  const countBadge = $(`a[data-test-row-filter="${filter}"] > div > .pf-c-badge`);
+  const count = await countBadge.getText();
+  await click($(`#${filter}`));
+  const rowsCount = await resourceRows.count();
+  expect(rowsCount).toBe(Number(count));
+};
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:30]: These functions can go inside the diskList object as well.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 7 above)
Comment (bipuladh) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:60]: https://github.com/openshift/console/pull/6606/files#diff-1e194e3ba92229ea049eea86f7bf2e94R47
If all the headers(columns) are visible I am confused why exactly some of the elements are missing here.

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+import { clickNavLink } from '@console/internal-integration-tests/views/sidenav.view';
+import { click } from '@console/shared/src/test-utils/utils';
+import {
+  page,
+  diskList,
+  clickFilterDropdown,
+  checkDiskFilter,
+  clearDiskFilter,
+} from '../views/disks-list-page.views';
+import { LOCAL_STORAGE_NAMESPACE } from '../../src/constants/index';
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:14]:
```suggestion
import { LOCAL_STORAGE_NAMESPACE } from '../../src/constants';
```

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts (hunk 14 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:17]: Hey,
The properties could run into self-referencing errors on moving inside object.

Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:17]: Its not harm keeping outside too.

File: frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts (hunk 15 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/views/disks-list-page.views.ts:30]: https://github.com/openshift/console/pull/6606#discussion_r490740481

File: frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts (hunk 7 above)
Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:60]: for headers the presence of DOM node is checked by the attribute data-label=<header-name>.
We are not using protractor get() and getText() functions which work on element type.

Comment (afreen23) [reply frontend/packages/local-storage-operator-plugin/integration-tests/tests/disks-list-page.scenario.ts:63]: I have posted the screenshot of running tests in description as well.
https://github.com/openshift/console/pull/6606#issue-486324168

**Review by bipuladh — APPROVED** [review]
/lgtm
